# Исследование надежности заемщиков.

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Шаг-1.-Обзор-данных" data-toc-modified-id="Шаг-1.-Обзор-данных"><span class="toc-item-num">1.&nbsp;&nbsp;</span>Обзор данных</a></span><li><span><a href="#Шаг-2.1-Заполнение-пропусков" data-toc-modified-id="Шаг-2.1-Заполнение-пропусков"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Заполнение пропусков</a></span></li><li><span><a href="#Шаг-2.2-Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Шаг-2.2-Проверка-данных-на-аномалии-и-исправления."><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Проверка данных на аномалии и исправления.</a></span></li><li><span><a href="#Шаг-2.3.-Изменение-типов-данных." data-toc-modified-id="Шаг-2.3.-Изменение-типов-данных."><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Изменение типов данных.</a></span></li><li><span><a href="#Шаг-2.4.-Удаление-дубликатов." data-toc-modified-id="Шаг-2.4.-Удаление-дубликатов."><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Удаление дубликатов.</a></span></li><li><span><a href="#Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма."><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Шаг-2.6.-Категоризация-дохода." data-toc-modified-id="Шаг-2.6.-Категоризация-дохода."><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Категоризация дохода.</a></span></li><li><span><a href="#Шаг-2.7.-Категоризация-целей-кредита." data-toc-modified-id="Шаг-2.7.-Категоризация-целей-кредита."><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Категоризация целей кредита.</a></span></li><li><span><a href="#Проверка-влияния-различных-факторов-на-вероятность-невозврата-кредита-в-срок." data-toc-modified-id="Проверка-влияния-различных-факторов-на-вероятность-невозврата-кредита-в-срок."><span class="toc-item-num">3.&nbsp;&nbsp;</span>Проверка влияния различных факторов на вероятность невозврата кредита в срок.</a></span></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-5"><span class="toc-item-num">4.&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

### Шаг 1. Обзор данных

In [110]:
import pandas as pd
import numpy as np

In [111]:
df = pd.read_csv('/datasets/data.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [112]:
display(df.head(10))

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

### Шаг 2.1 Заполнение пропусков

In [113]:
display(df.loc[df['days_employed'].isna()].head(15))
#Обнаружил значения NaN, что означает принадлежность к типу float. Видно что некоторые одинаковые значения написаны по разному в столбце education.

children  days_employed  dob_years education  education_id  \
12          0            NaN         65   среднее             1   
26          0            NaN         41   среднее             1   
29          0            NaN         63   среднее             1   
41          0            NaN         50   среднее             1   
55          0            NaN         54   среднее             1   
65          0            NaN         21   среднее             1   
67          0            NaN         52    высшее             0   
72          1            NaN         32    высшее             0   
82          2            NaN         50    высшее             0   
83          0            NaN         52   среднее             1   
90          2            NaN         35    высшее             0   
94          1            NaN         34    высшее             0   
96          0            NaN         44   СРЕДНЕЕ             1   
97          0            NaN         47    высшее             0   
120         0            NaN         46   среднее             1   

             family_status  family_status_id gender  income_type  debt  \
12        гражданский брак                 1      M    пенсионер     0   
26         женат / замужем                 0      M  госслужащий     0   
29   Не женат / не замужем                 4      F    пенсионер     0   
41         женат / замужем                 0      F  госслужащий     0   
55        гражданский брак                 1      F    пенсионер     1   
65   Не женат / не замужем                 4      M    компаньон     0   
67         женат / замужем                 0      F    пенсионер     0   
72         женат / замужем                 0      M  госслужащий     0   
82         женат / замужем                 0      F    сотрудник     0   
83         женат / замужем                 0      M    сотрудник     0   
90         женат / замужем                 0      F    сотрудник     0   
94        гражданский брак                 1      F    компаньон     0   
96         женат / замужем                 0      F    сотрудник     0   
97         женат / замужем                 0      F    сотрудник     0   
120        женат / замужем                 0      F    сотрудник     0   

     total_income                                purpose  
12            NaN                        сыграть свадьбу  
26            NaN                            образование  
29            NaN       строительство жилой недвижимости  
41            NaN       сделка с подержанным автомобилем  
55            NaN                        сыграть свадьбу  
65            NaN  операции с коммерческой недвижимостью  
67            NaN                покупка жилья для семьи  
72            NaN  операции с коммерческой недвижимостью  
82            NaN                                  жилье  
83            NaN                                  жилье  
90            NaN                      операции с жильем  
94            NaN                  на проведение свадьбы  
96            NaN             покупка жилой недвижимости  
97            NaN                 профильное образование  
120           NaN                     высшее образование

In [114]:
#Доля пропущенных значений в каждем столбце
#print(df['days_employed'].isna().count()/df['days_employed'].count())#закомментировал как считал раньше этот показатель в days_employed и total_income
#print(df['total_income'].isna().count()/df['total_income'].count())
print(df['days_employed'].isna().sum())#вывод кол-ва пропусков
print(df['days_employed'].count())#кол-во строк без пропусков
print(len(df))#общее число строк в датасете
print()
print(df['days_employed'].isna().sum()/len(df))#заново посчитал долю пропущенных значений в каждом столбце
print(df['total_income'].isna().sum()/len(df))

2174
19351
21525

0.10099883855981417
0.10099883855981417


Доли пропущенных значений в столбцах days_employed и total_income совпадают.
Возможные причины пропусков данных:
- Потеря данных при переносе
- Человек находится в поиске работы или недавно сменил место работы
- Человек на пенсии и не имеет место работы

Удалить данные с пропусками в данном случае является не лучшей идеей, так как они составляют значительную долю (более 10%) и могут содержать полезную информацию. Их лучше заменить медианным значением, так как это окажет минимальное воздействие на будущий результат. При использовании среднего возможно отклонение результата в сторону выбросов (слишком малых или слишком больших значений).

In [115]:
df['days_employed'] = abs(df['days_employed'])#заменил данные по модулю для верного расчета медианы
df['days_employed'] = df['days_employed'].fillna(df['days_employed'].median())
df['total_income'] = df['total_income'].fillna(df['total_income'].median())
display(df.head(15))

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42               высшее             0   
1          1    4024.803754         36              среднее             1   
2          0    5623.422610         33              Среднее             1   
3          3    4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0     926.185831         27               высшее             0   
6          0    2879.202052         43               высшее             0   
7          0     152.779569         50              СРЕДНЕЕ             1   
8          2    6929.865299         35               ВЫСШЕЕ             0   
9          0    2188.756445         41              среднее             1   
10         2    4171.483647         36               высшее             0   
11         0     792.701887         40              среднее             1   
12         0    2194.220567         65              среднее             1   
13         0    1846.641941         54  неоконченное высшее             2   
14         0    1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12  145017.937533                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

Возможно отрицательные значения в столбце days_employed появились в результате неверного форматирования даты при переводе в численный формат.
Далее заменим данные в столбце total_income на целочисленный формат.

### Шаг 2.2 Проверка данных на аномалии и исправления.

Необычные значения есть в столбцах с количеством детей (-1, 20), возрасте (0) и гендере (XNA). Эти значения можно обнаружить используя метод unique() или составить сводную таблицу. Заменил выбросы в столбцах с детьми, возрасте и гендере на None (NaN), чтобы не возникало проблем при расчётах.

In [116]:
#display(df.loc[df['children'] == -1].head())# для проверки значений в строках (вывод где значение детей -1)
df.loc[df['children'] == -1 ,'children'] = 1 # меняем -1 ребёнка на +1 
print('Строки с аномалиями, которые удаляем:', df.query("dob_years == 0 | gender == 'XNA' | children == 20")['children'].count()) #Да значений с 20 детьми, неизвестным гендером и нулевым возрастом не так много (менее 1%), поэтому их можно удалить
print(f'Сколько было строк: {len(df)}')
df = df.query("dob_years != 0 and gender != 'XNA' and children != 20")# ещё можно удалить строки через использование drop() и loc
print(f'Сколько стало строк: {len(df)}')
#display(df.loc[df['children'].isna()].head())# вывод таблицы со значениями NaN

Строки с аномалиями, которые удаляем: 177
Сколько было строк: 21525
Сколько стало строк: 21348


### Шаг 2.3. Изменение типов данных.

In [117]:
df['total_income'] = df['total_income'].astype('int')
display(df.head())

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

### Шаг 2.4. Удаление дубликатов.

In [118]:
print(df.duplicated().sum())
display(df.loc[df.duplicated()].head())#дубликаты появились в строках, где были пропущенные значения.

54


children  days_employed  dob_years education  education_id  \
2849         0    2194.220567         41   среднее             1   
4182         1    2194.220567         34    ВЫСШЕЕ             0   
4851         0    2194.220567         60   среднее             1   
5557         0    2194.220567         58   среднее             1   
7808         0    2194.220567         57   среднее             1   

         family_status  family_status_id gender income_type  debt  \
2849   женат / замужем                 0      F   сотрудник     0   
4182  гражданский брак                 1      F   сотрудник     0   
4851  гражданский брак                 1      F   пенсионер     0   
5557  гражданский брак                 1      F   пенсионер     0   
7808  гражданский брак                 1      F   пенсионер     0   

      total_income                  purpose  
2849        145017  покупка жилья для семьи  
4182        145017                  свадьба  
4851        145017                  свадьба  
5557        145017          сыграть свадьбу  
7808        145017    на проведение свадьбы

In [119]:
df = df.drop_duplicates().reset_index(drop=True)
display(df.tail())#количество строк уменьшилось на кол-во дубликатов

children  days_employed  dob_years education  education_id  \
21289         1    4529.316663         43   среднее             1   
21290         0  343937.404131         67   среднее             1   
21291         1    2113.346888         38   среднее             1   
21292         3    3112.481705         38   среднее             1   
21293         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
21289  гражданский брак                 1      F   компаньон     0   
21290   женат / замужем                 0      F   пенсионер     0   
21291  гражданский брак                 1      M   сотрудник     1   
21292   женат / замужем                 0      M   сотрудник     1   
21293   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose  
21289        224791             операции с жильем  
21290        155999          сделка с автомобилем  
21291         89672                  недвижимость  
21292        244093  на покупку своего автомобиля  
21293         82047         на покупку автомобиля

Ниже закомментировал вариант, где устранил неявные дубликаты в столбце purpose с помощью метода replace(). Это удобно при небольшом наборе целевых категорий. При бОльшем количестве удобнее искать по ключевым словам, что и сделал в шаге 2.7.

In [120]:
print(df['education'].unique())
df['education'] = df['education'].str.lower()
print(df['education'].unique())

#df['purpose'] = df['purpose'].replace(['покупка жилья', 'операции с жильем', 'покупка жилья для семьи', 'покупка недвижимости', 'покупка коммерческой недвижимости', 'покупка жилой недвижимости', 'строительство собственной недвижимости', 'недвижимость', 'строительство недвижимости', 'операции с коммерческой недвижимостью', 'строительство жилой недвижимости', 'жилье',
# 'операции со своей недвижимостью', 'покупка своего жилья', 'покупка жилья для сдачи', 'ремонт жилью'], 'операции с недвижимостью')
#df['purpose'] = df['purpose'].replace(['приобретение автомобиля', 'на покупку подержанного автомобиля', 'на покупку своего автомобиля', 'автомобили', 'сделка с подержанным автомобилем', 'автомобиль', 'свой автомобиль', 'сделка с автомобилем', 'на покупку автомобиля'], 'операции с автомобилем')
#df['purpose'] = df['purpose'].replace(['дополнительное образование', 'образование', 'заняться образованием', 'получение дополнительного образования', 'получение высшего образования', 'профильное образование', 'высшее образование', 'заняться высшим образованием'], 'получение образования')
#df['purpose'] = df['purpose'].replace(['сыграть свадьбу', 'на проведение свадьбы', 'свадьба'], 'проведение свадьбы')
#print(df['purpose'].unique())#заменил неявные дубликаты в столбце propose
#print(df.duplicated().sum())

['высшее' 'среднее' 'Среднее' 'СРЕДНЕЕ' 'ВЫСШЕЕ' 'неоконченное высшее'
 'начальное' 'Высшее' 'НЕОКОНЧЕННОЕ ВЫСШЕЕ' 'Неоконченное высшее'
 'НАЧАЛЬНОЕ' 'Начальное' 'Ученая степень' 'УЧЕНАЯ СТЕПЕНЬ'
 'ученая степень']
['высшее' 'среднее' 'неоконченное высшее' 'начальное' 'ученая степень']


In [121]:
try:
    df = df.drop_duplicates().reset_index(drop=True)#сделал эту конструкцию, если вдруг будет ошибка при отсутствии дубликатов
    print('Повторяющихся строк после удаления явных и неявных дубликатов:')
    print(df.duplicated().sum())
except:
    print('Дубликатов нет')

Повторяющихся строк после удаления явных и неявных дубликатов:
0


Явные дубликаты нашёл методом duplicated() и удалил их drop_duplicates().reset_index(drop=True). В столбце education привёл данные к единому регистру, а затем как вариант можно привести неявные дубликаты в столбце purpose к 4 основным типам: операции с недвижимостью, операции с автомобилем, проведение свадьбы и получение образования. Использовал для этого метод replace(). Можно этого не делать, т.к. в дальнейшем мы проведём категоризацию по цели, что удобнее.
Возможные причины появления дубликатов:
- Один человек подавал несколько заявок.
- Техническая ошибка при заполнении заявки, переносе данных.
- Человеческий фактор (при составлении анкеты, консультант заполнил её несколько раз).

### Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

In [122]:
df_edu = df[['education','education_id']]
df_edu = df_edu.drop_duplicates().reset_index(drop=True)
display(df_edu)#словарь для образования

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

In [123]:
df_fam_status = df[['family_status','family_status_id']]
df_fam_status = df_fam_status.drop_duplicates().reset_index(drop=True)
display(df_fam_status)#словарь для семейного положения

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  Не женат / не замужем                 4

In [124]:
df.drop(columns = ['education','family_status'],axis = 1, inplace=True)#удаляем столбцы education, family_status
display(df.head())

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0  340266.072047         53             1                 1      F   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу

### Шаг 2.6. Категоризация дохода.

In [125]:
def income_category(income):
    if income <= 30000:
        return 'E'
    elif income <= 50000:
        return 'D'
    elif income <= 200000:
        return 'C'
    elif income <= 1000000:
        return 'B'
    else:
        return 'A'
print(income_category(1115000))#проверка
df['total_income_category'] = df['total_income'].apply(income_category)
display(df.head())

A


children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0  340266.072047         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C

### Шаг 2.7. Категоризация целей кредита.

In [126]:
def purpose_identification(string):
         if 'авто' in string:
            return 'операции с автомобилем'
         if 'недвиж' in string or 'жиль' in string:
            return 'операции с недвижимостью'
         if 'образ' in string:
            return 'получение образования'
         if 'свадьб' in string:
            return 'проведение свадьбы'
print(purpose_identification('на свадьбу'))
df['purpose_category'] = df['purpose'].apply(purpose_identification)
display(df.head(15))
print(df['purpose_category'].unique())

проведение свадьбы


children  days_employed  dob_years  education_id  family_status_id gender  \
0          1    8437.673028         42             0                 0      F   
1          1    4024.803754         36             1                 0      F   
2          0    5623.422610         33             1                 0      M   
3          3    4124.747207         32             1                 0      M   
4          0  340266.072047         53             1                 1      F   
5          0     926.185831         27             0                 1      M   
6          0    2879.202052         43             0                 0      F   
7          0     152.779569         50             1                 0      M   
8          2    6929.865299         35             0                 1      F   
9          0    2188.756445         41             1                 0      M   
10         2    4171.483647         36             0                 0      M   
11         0     792.701887         40             1                 0      F   
12         0    2194.220567         65             1                 1      M   
13         0    1846.641941         54             2                 0      F   
14         0    1844.956182         56             0                 1      F   

   income_type  debt  total_income                            purpose  \
0    сотрудник     0        253875                      покупка жилья   
1    сотрудник     0        112080            приобретение автомобиля   
2    сотрудник     0        145885                      покупка жилья   
3    сотрудник     0        267628         дополнительное образование   
4    пенсионер     0        158616                    сыграть свадьбу   
5    компаньон     0        255763                      покупка жилья   
6    компаньон     0        240525                  операции с жильем   
7    сотрудник     0        135823                        образование   
8    сотрудник     0         95856              на проведение свадьбы   
9    сотрудник     0        144425            покупка жилья для семьи   
10   компаньон     0        113943               покупка недвижимости   
11   сотрудник     0         77069  покупка коммерческой недвижимости   
12   пенсионер     0        145017                    сыграть свадьбу   
13   сотрудник     0        130458            приобретение автомобиля   
14   компаньон     1        165127         покупка жилой недвижимости   

   total_income_category          purpose_category  
0                      B  операции с недвижимостью  
1                      C    операции с автомобилем  
2                      C  операции с недвижимостью  
3                      B     получение образования  
4                      C        проведение свадьбы  
5                      B  операции с недвижимостью  
6                      B  операции с недвижимостью  
7                      C     получение образования  
8                      C        проведение свадьбы  
9                      C  операции с недвижимостью  
10                     C  операции с недвижимостью  
11                     C  операции с недвижимостью  
12                     C        проведение свадьбы  
13                     C    операции с автомобилем  
14                     C  операции с недвижимостью

['операции с недвижимостью' 'операции с автомобилем'
 'получение образования' 'проведение свадьбы']


### Проверка влияния различных факторов на вероятность невозврата кредита в срок.

При составлении сводной таблицы заметил заметил необычные значения в количестве детей (-1, 20), возрасте (0) и гендере (XNA). Заменим выбросы в столбцах с детьми и возрасте на None (NaN), чтобы не возникало проблем при расчётах. Значение в гендере можно сначала уточнить в документации или у сотрудников, которые занимались выгрузкой, а потом либо заменить на None, либо оставтить как есть.

In [127]:
#display(df.pivot_table(index='children', values='debt', aggfunc='mean'))# попробовал составить сводную таблицу с детьми и заметил выбросы -1 и 20. Количество детей "20" встречается в нескольких строках, что даёт сделать вывод о выбросе, а не о реальном значении.
#display(df.loc[df['children'] == -1].head())# проверил данные в строках, где значение детей -1
#df.loc[df['children'] == -1 ,'children'] = None
#df.loc[df['children'] == 20 ,'children'] = None
#df.loc[df['dob_years'] == 0 ,'dob_years'] = None
#display(df.loc[df['children'].isna()].head())# вывод таблицы со значениями NaN

In [128]:
display(df.pivot_table(index='children', values='debt', aggfunc='mean'))

debt
children          
0         0.075458
1         0.091341
2         0.095145
3         0.082317
4         0.097561
5         0.000000

Однозначный вывод о зависисмости количества детей и возврата кредита в срок сделать нельзя, так как рост вероятности невозврата в срок растёт только до первых двух детей, затем резко падаёт на 3, увеличивается на 4 и отсутствует у тех, у кого 5 детей (выводы на основе исследуемых данных).

In [129]:
pivot_table_fs = df.pivot_table(index='family_status_id', values='debt', aggfunc='mean')
pivot_table_fs = pivot_table_fs.merge(df_fam_status, on = 'family_status_id', how='left')
display(pivot_table_fs)

family_status_id      debt          family_status
0                 0  0.075478        женат / замужем
1                 1  0.093029       гражданский брак
2                 2  0.065263         вдовец / вдова
3                 3  0.071006              в разводе
4                 4  0.097666  Не женат / не замужем

Здесь более понятные результаты. Так, видно, что наибольшая вероятность невозврата кредита в срок у людей в гражданском браке и у неженатых/незамужних. Это можно объяснить тем, что люди не вступившие в брак чаще всего молоды и не обрели опыта для устройства на более высокооплачиваемую работу и ещё не накопили достаточную "подушку безопасности". В гражданском браке люди рискуют только своим личным имуществом, а не имуществом семьи. В браке в случае сложной ситуации больше вероятность, что отдать кредит может помочь партнеёр. Для подтверждения своей гипотезы составил сводную таблицу со средним возрастом заемщика в зависимости от семейного положения.

Самая высокая вероятность вернуть кредит в срок у вдов и вдовцов. При этом эта категория является самой возрастной (средний возраст 57 лет).

In [130]:
age_on_fs = df.pivot_table(index='family_status_id', values='dob_years', aggfunc='mean')# сводная таблица со средним возрастом по категориям семейного положения
display(age_on_fs.merge(df_fam_status, on = 'family_status_id', how='left'))

family_status_id  dob_years          family_status
0                 0  43.727414        женат / замужем
1                 1  42.277629       гражданский брак
2                 2  56.782105         вдовец / вдова
3                 3  45.908707              в разводе
4                 4  38.606104  Не женат / не замужем

Нельзя сделать однозначный вывод о зависимости уровня доходов и вероятности невозврата кредита в срок. Если рассмотреть категорию D (30 - 50 тыс.), то видно что вероятность невозврата кредита в срок выше как в категории с более высоким уровнем доходов С (50 - 200 тыс.), так и в категории с более низким уровнем доходов E (менее 30 тыс.).

In [131]:
display(df.pivot_table(index='total_income_category', values='debt', aggfunc='mean'))

debt
total_income_category          
A                      0.080000
B                      0.070685
C                      0.084781
D                      0.060345
E                      0.090909

Самая высокая вероятность невозврата кредита в срок у целей на получение образования и операций с автомобирем. Самая низкая вероятность невозврата у операций с недвижимостью - здесь люди рискуют своей "крышей над головой". Проведение свадьбы занимает промежуточное положение.

In [132]:
display(df.pivot_table(index='purpose_category', values='debt', aggfunc='mean'))

debt
purpose_category                  
операции с автомобилем    0.093230
операции с недвижимостью  0.072434
получение образования     0.092714
проведение свадьбы        0.078661

Попробуем проверить более сложную зависимость: цель и доход заемщика с вероятностью возврата кредита в срок.

In [133]:
display(df.pivot_table(index= ['purpose_category','total_income_category'], values='debt', aggfunc='mean'))

debt
purpose_category         total_income_category          
операции с автомобилем   A                      0.000000
                         B                      0.083333
                         C                      0.095463
                         D                      0.140845
                         E                      0.000000
операции с недвижимостью A                      0.058824
                         B                      0.067138
                         C                      0.074765
                         D                      0.035928
                         E                      0.181818
получение образования    A                      0.250000
                         B                      0.077966
                         C                      0.098671
                         D                      0.026316
                         E                      0.000000
проведение свадьбы       A                      0.000000
                         B                      0.051661
                         C                      0.087260
                         D                      0.088235
                         E                      0.000000

Виден тренд роста вероятности невозврата кредита в срок с понижением дохода в категориях: операции с автомобилем, операции с недвижимостью, проведение свадьбы. Слишком высокая вероятность невозврата кредита в срок у самой обеспеченной группы (А - доход выше 1 млн) при получении денег на образование. Возможно это связано с низкой представленностью данной группы в исследовании (репрезентативность выборки), поэтому получается такой результат. 
Возможно кто-то брал кредит на обучение в Оксфорде или Гарварде:)

## Общий вывод:

Явной зависимости между возвратом кредита в срок и количеством детей нет. Также сложно проследить зависимость между уровнем дохода и возвратом кредита в срок. Проблемы с возвратом кредита могут быть как у более обеспеченной категории с ежемесячным доходом 50-200 тыс., так и у категории с низким ежемесячным уровнем доходом до 30 тыс. При этом вероятность возврата кредита в срок у категории 30-50 тыс. выше, чем у двух перечисленных.
На возврат кредита в срок влияют семеное положение и цель кредита. Ниже веротность вернуть кредит в срок у неженатой категории людей и у тех, кто берёт деньги на покупку автомобиля или получение образования. Выше вероятность вернуть долг в срок у людей, состоящий в браке, и вдов (вдовцов). Это объясняется тем, что эти категории старше, имеют больший опыт работы и "подушку безопасности".
Высокая вероятность вернуть кредит в срок у людей, которые берут его на покупку недвижимости, так как они рискуют своей "крышей над головой".

Внёс правки в аномалиях: заменил -1 на 1. Строки, где значение нельзя исправить на подходящие, удалил. Они составляли менее 1% (177 строк).

In [134]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21277 entries, 0 to 21276
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   children               21277 non-null  int64  
 1   days_employed          21277 non-null  float64
 2   dob_years              21277 non-null  int64  
 3   education_id           21277 non-null  int64  
 4   family_status_id       21277 non-null  int64  
 5   gender                 21277 non-null  object 
 6   income_type            21277 non-null  object 
 7   debt                   21277 non-null  int64  
 8   total_income           21277 non-null  int64  
 9   purpose                21277 non-null  object 
 10  total_income_category  21277 non-null  object 
 11  purpose_category       21277 non-null  object 
dtypes: float64(1), int64(6), object(5)
memory usage: 1.9+ MB
